pkg/serviceaccounts/oauthclient/oauthclientregistry.go

```diff
@@ -73,14 +203,117 @@ func (a *saOAuthClientAdapter) GetClient(ctx kapi.Context, name string) (*oautha
 		// TODO update this to allow https redirection to any
 		// 1. service IP (useless in general)
 		// 2. service DNS (useless in general)
-		// 3. route DNS (useful)
-		// 4. loopback? (useful, but maybe a bit weird)
-		RedirectURIs: redirectURIs,
+		// 3. loopback? (useful, but maybe a bit weird)
+		RedirectURIs: sets.NewString(redirectURIs...).List(),
 		GrantMethod:  a.grantMethod,
 	}
 	return saClient, nil
 }
 
+func parseModelsMap(annotations map[string]string) map[string]model {
+	models := map[string]model{}
+	for key, value := range annotations {
+		prefix, name, ok := parseModelPrefixName(key)
+		if !ok {
+			continue
+		}
+		m := models[name]
+		switch prefix {
+		case OAuthRedirectModelAnnotationURISchemePrefix:
+			m.scheme = value
+		case OAuthRedirectModelAnnotationURIPortPrefix:
+			m.port = value
+		case OAuthRedirectModelAnnotationURIPathPrefix:
+			m.path = value
+		case OAuthRedirectModelAnnotationURIHostPrefix:
+			m.host = value
+		case OAuthRedirectModelAnnotationResourceKindPrefix:
+			m.kind = value
+		case OAuthRedirectModelAnnotationResourceNamePrefix:
+			m.name = value
+		case OAuthRedirectModelAnnotationResourceGroupPrefix:
+			m.group = value
+		}
+		models[name] = m
+	}
+	return models
+}
+
+func parseModelPrefixName(key string) (string, string, bool) {
+	for _, prefix := range modelPrefixes {
+		if strings.HasPrefix(key, prefix) {
+			return prefix, key[len(prefix):], true
+		}
+	}
+	return "", "", false
+}
+
+func (a *saOAuthClientAdapter) extractRedirectURIs(modelsMap map[string]model, namespace string) redirectURIList {
+	var data redirectURIList
+	groupKindModelListMapper := map[unversioned.GroupKind]modelList{}
+	groupKindModelToURI := map[unversioned.GroupKind]namesToObjMapperFunc{
+		routeGroupKind: a.redirectURIsFromRoutes,
+	}
+
+	for _, model := range modelsMap {
+		gk := model.getGroupKind()
+		if _, ok := groupKindModelToURI[gk]; ok {
+			groupKindModelListMapper[gk] = append(groupKindModelListMapper[gk], model)
+		}
+	}
+
+	for gk, models := range groupKindModelListMapper {
+		if len(models) > 0 {
+			if names := models.getNames(); names.Len() > 0 {
+				if objMapper := groupKindModelToURI[gk](namespace, names); len(objMapper) > 0 {
+					data = append(data, models.getRedirectURIs(objMapper)...)
+				}
+			}
+		}
+	}
+
+	return data
+}
+
+func (a *saOAuthClientAdapter) redirectURIsFromRoutes(namespace string, osRouteNames sets.String) map[string]redirectURIList {
+	var routes []routeapi.Route
+	routeInterface := a.routeClient.Routes(namespace)
+	if osRouteNames.Len() > 1 {
+		if r, err := routeInterface.List(kapi.ListOptions{}); err == nil {
+			routes = r.Items
+		}
+	} else {
+		if r, err := routeInterface.Get(osRouteNames.List()[0]); err == nil {
```

Comment:
`getNames` is used to build the `sets.String` and it filters out empty strings.  Thus at this point in the code we are guaranteed to have at least one item in the set and all items will be non-empty strings.